### PR TITLE
Refactor commitments and cleanup

### PR DESCRIPTION
**Pull request by CjS77**
Refactor commitments to allow serialisation (and generally be better)
    
Our CommitmentFactory struct was a bit of a hack TBH.

This splits out stuff so that it's a proper factory, and we pass
 a factory instance around to functions that need CommitmentFactory
 services.
    
Commitments can be serialised now, because the reference to `base` has been dropped.
    
The API is generally friendlier.
    
## Notes on Serialisation:
    
The Dalek libraries only serialize to binary (understandably), but this has 2 yucky implications:
    
1. Exporting to "human readable" formats like JSON yield crappy looking 'binary arrays', e.g. /[12, 223, 65, .../]
2. Reading back from JSON is broken because serde doesn't read this back as a byte string, but as a seq.
    
The workaround is to have binary serialization by default, but if a struct is going to be saved in JSON format, then you can override that behaviour with `with_serialize`, e.g.
    
    ```rust
      #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
      pub struct KeyManager<K: SecretKey, D: Digest> {
          #[serde(serialize_with = "serialise_to_hex", deserialize_with = "secret_from_hex")]
          pub master_key: K,
          pub branch_seed: String,
          pub primary_key_index: usize,
          digest_type: PhantomData<D>,
      }
    ```
    
## Breaking changes:
    
1. Commitments are now just Pubkey Newtypes. This means that we don't check that the bases are equal anymore.
       In practice, I don't think this changes anything.

### CODE REVIEW

File: base_layer/core/src/transaction_protocol/sender.rs [hunk 1]
```diff
@@ -262,7 +268,7 @@ impl SenderTransactionProtocol {
             .with_signature(&s_agg)
             .build()?;
         tx_builder.with_kernel(kernel);
-        let tx = tx_builder.build()?;
+        let tx = tx_builder.build(prover, factory)?;
```
Comment (neonknight64) [inline base_layer/core/src/transaction_protocol/sender.rs]: nit: ln 271 and 272 can be combined  as only `tx_builder.build(prover, factory)`

File: base_layer/keymanager/src/keymanager.rs [hunk 2]
```diff
@@ -48,6 +51,7 @@ where K: SecretKey
 
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
 pub struct KeyManager<K: SecretKey, D: Digest> {
+    #[serde(serialize_with = "serialize_to_hex", deserialize_with = "secret_from_hex")]
```
Comment (neonknight64) [inline base_layer/keymanager/src/keymanager.rs]: This is a nice new feature and will be very handy

File: base_layer/core/src/transaction_protocol/sender.rs (hunk 1 above)
Comment (CjS77) [reply base_layer/core/src/transaction_protocol/sender.rs]: Good catch, thanks

Comment (CjS77) [reply base_layer/core/src/transaction_protocol/sender.rs]: Fixed.

**Pull request merged**